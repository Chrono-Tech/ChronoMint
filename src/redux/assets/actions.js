import contractsManagerDAO from 'dao/ContractsManagerDAO'

export const ASSETS_LIST = 'assets/LIST'

export const listAssets = () => async (dispatch) => {
  const dao = await contractsManagerDAO.getAssetsManagerDAO()
  dispatch({type: ASSETS_LIST, list: await dao.getList()})
}
