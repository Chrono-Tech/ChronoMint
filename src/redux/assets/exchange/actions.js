import type ExchangeManagerDAO from 'dao/ExchangeManagerDAO'
import type ExchangeModel from 'models/ExchangeModel'
import type TokenModel from 'models/TokenModel'
import contractsManagerDAO from 'dao/ContractsManagerDAO'

export const EXCHANGE_MANAGER_LIST = 'exchangeManager/LIST'
export const EXCHANGE_MANAGER_SET = 'exchangeManager/SET'
export const EXCHANGE_MANAGER_REMOVE = 'exchangeManager/REMOVE'

const getDAO = async (): Promise<ExchangeManagerDAO> => await contractsManagerDAO.getExchangeManagerDAO()

export const listExchanges = () => async (dispatch) => {
  const dao = await getDAO()
  dispatch({type: EXCHANGE_MANAGER_LIST, list: await dao.getList()})
}

export const addExchange = (address) => async () => {
  const exchange = new ExchangeModel({address, isFetching: true})
  dispatch({type: EXCHANGE_MANAGER_SET, exchange})
  try {
    const dao = await getDAO()
    await dao.addExchange(address)
  } catch (e) {
    // no rollback
  }
}

export const createExchange = (token: TokenModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.createExchange(token)
  } catch (e) {
    // no rollback
  }
}

export const editExchange = (exchange: ExchangeModel, newExchange: ExchangeModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.editExchange(exchange, newExchange)
  } catch (e) {
    // no rollback
  }
}

export const removeExchange = (exchange: ExchangeModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.removeExchange(exchange)
  } catch (e) {
    // no rollback
  }
}

// TODO @bshevchenko: MINT-230 AssetsManager
// export const addOwner = (exchange: ExchangeModel, account) => async () => {
//   try {
//     const dao = await getDAO()
//     await dao.addOwner(exchange, account)
//   } catch (e) {
//     // no rollback
//   }
// }
//
// export const removeOwner = (exchange: ExchangeModel, account) => async () => {
//   try {
//     const dao = await getDAO()
//     await dao.removeOwner(exchange, account)
//   } catch (e) {
//     // no rollback
//   }
// }
