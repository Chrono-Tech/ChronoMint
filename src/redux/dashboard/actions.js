import LHTProxyDAO from '../../dao/LHTProxyDAO'
import UserStorageDAO from '../../dao/UserStorageDAO'

import {
  DASHBOARD_TOTAL_LHT_FETCH,
  DASHBOARD_TOTAL_LHT,
  DASHBOARD_TOTAL_MEMBERS_FETCH,
  DASHBOARD_TOTAL_MEMBERS
} from './reducer'

export const updateTotalLHT = () => (dispatch) => {
  dispatch({type: DASHBOARD_TOTAL_LHT_FETCH})
  return LHTProxyDAO.totalSupply()
    .then(balance => {
      dispatch({type: DASHBOARD_TOTAL_LHT, payload: balance})
    })
}

export const updateTotalMembers = () => (dispatch) => {
  dispatch({type: DASHBOARD_TOTAL_MEMBERS_FETCH})
  return UserStorageDAO.usersTotal()
    .then(number => {
      dispatch({type: DASHBOARD_TOTAL_MEMBERS, payload: number})
    })
}
