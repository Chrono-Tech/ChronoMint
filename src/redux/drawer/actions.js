/**
 * Copyright 2017â€“2018, LaborX PTY
 * Licensed under the AGPL Version 3 license.
 */

import {
  DRAWER_HIDE,
  DRAWER_TOGGLE,
} from './constants'

export const drawerToggle = () => (dispatch) => {
  dispatch({ type: DRAWER_TOGGLE })
}
export const drawerHide = () => (dispatch) => {
  dispatch({ type: DRAWER_HIDE })
}
