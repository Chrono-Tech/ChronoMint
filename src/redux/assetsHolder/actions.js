import { DUCK_SESSION } from '@/redux/session/actions'
import contractsManagerDAO from 'dao/ContractsManagerDAO'
import TIMEHolderDAO from 'dao/TIMEHolderDAO'
import Amount from 'models/Amount'
import ApprovalNoticeModel from 'models/notices/ApprovalNoticeModel'
import TokenModel from 'models/tokens/TokenModel'
import { DUCK_TOKENS } from 'redux/tokens/actions'
import tokenService, { EVENT_NEW_TOKEN } from 'services/TokenService'

export const DUCK_ASSETS_HOLDER = 'assetsHolder'

export const TIME_HOLDER_INIT = 'timeHolder/INIT'
export const TIME_HOLDER_TIME_ADDRESS = 'timeHolder/timeAddress'
export const TIME_HOLDER_WALLET_ADDRESS = 'timeHolder/timeHolderWalletAddress'
export const TIME_HOLDER_ADDRESS = 'timeHolder/timeHolderAddress'

let timeHolderDAO: TIMEHolderDAO = null

const subscribeOnTokens = (token: TokenModel) => async (dispatch, getState) => {
  if (token.symbol() !== 'TIME') {
    return
  }

  dispatch({ type: TIME_HOLDER_TIME_ADDRESS, address: token.address() })

  // init
  const [ timeHolderAddress, timeHolderWalletAddress ] = await Promise.all([
    timeHolderDAO.getAddress(),
    timeHolderDAO.getWalletAddress(),
  ])

  dispatch({ type: TIME_HOLDER_ADDRESS, address: timeHolderAddress })
  dispatch({ type: TIME_HOLDER_WALLET_ADDRESS, address: timeHolderWalletAddress })

  // TODO @dkchv: review again
  const contractNames = {}
  contractNames[ timeHolderAddress ] = 'TIME Holder'
  ApprovalNoticeModel.setContractNames(contractNames)

  // update allowance
  const timeDAO = tokenService.getDAO(token.id())
  const { account } = getState().get(DUCK_SESSION)
  const [ timeHolderAllowance, timeHolderWalletAllowance ] = await Promise.all([
    timeDAO.getAccountAllowance(timeHolderAddress, account),
    timeDAO.getAccountAllowance(timeHolderWalletAddress, account),
  ])

  console.log('--assetHolder: ', +timeHolderAllowance, +timeHolderWalletAllowance)

  // token = token
  //   .setAllowance(timeHolderAddress, timeHolderAllowance)
  //   .setAllowance(timeHolderWalletAddress, timeHolderWalletAllowance)
}

export const initAssetsHolder = () => async (dispatch, getState) => {
  if (getState().get(DUCK_ASSETS_HOLDER).isInited()) {
    return
  }
  dispatch({ type: TIME_HOLDER_INIT, inInited: true })

  const { account } = getState().get(DUCK_SESSION)
  timeHolderDAO = await contractsManagerDAO.getTIMEHolderDAO()
  const [ assets, deposit ] = await Promise.all([
    timeHolderDAO.getSharesContract(),
    timeHolderDAO.getDeposit(account),
  ])
  console.log('--deposit', assets, +deposit)

  // subscribe to tokens
  const callback = (token) => dispatch(subscribeOnTokens(token))
  tokenService.on(EVENT_NEW_TOKEN, callback)
  // fetch for existing tokens
  const tokens = getState().get(DUCK_TOKENS)
  tokens.list().forEach(callback)
}

export const depositAsset = (amount: Amount, token: TokenModel) => async (dispatch) => {
  // const amountBN = new BigNumber(amount)

  // dispatch(balanceMinus(amountBN, token))
  const assetDAO = tokenService.getDAO(token)

  try {
    await timeHolderDAO.deposit(amount, assetDAO)
  } catch (e) {
    // eslint-disable-next-line
    console.error('deposit error', e.message)
  } finally {
    // compensation for update in watchTransfer
    // dispatch(balancePlus(amount, token))
  }
}

export const withdrawAsset = (amount: Amount, token: TokenModel) => async () => {
  // const amountBN = new BigNumber(amount)

  // dispatch(depositMinus(amountBN))
  const assetDAO = tokenService.getDAO(token)

  try {
    await timeHolderDAO.withdraw(amount, assetDAO)
  } catch (e) {
    // eslint-disable-next-line
    console.error('withdraw error', e.message)
  } finally {
    // dispatch(depositPlus(amountBN))
  }
}
