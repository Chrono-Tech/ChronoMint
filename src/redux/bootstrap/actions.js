import ls from 'utils/LocalStorage'
import { LOCAL_ID } from 'network/settings'

import {
  checkMetaMask, checkLocalSession, restoreLocalSession, createNetworkSession,
  checkTestRPC
} from '../network/actions'
import { login } from '../session/actions'

export const bootstrap = (relogin = true) => async (dispatch) => {
  dispatch(checkMetaMask())
  dispatch(checkTestRPC())

  if (!relogin) {
    return
  }

  const localAccount = ls.getLocalAccount()
  const isPassed = await dispatch(checkLocalSession(localAccount))
  if (isPassed) {
    await dispatch(restoreLocalSession(localAccount))
    dispatch(createNetworkSession(localAccount, LOCAL_ID, LOCAL_ID))
    dispatch(login(localAccount))
  } else {
    // eslint-disable-next-line
    console.warn('Can\'t restore local session')
  }
}
