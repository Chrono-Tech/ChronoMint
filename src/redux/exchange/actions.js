// TODO @bshevchenko: this is intermediate version for demo
import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import { ETH, LHT } from 'redux/wallet/actions'
import ExchangeOrderModel from 'models/ExchangeOrderModel'
import exchangeDAO from 'dao/ExchangeDAO'
import ethereumDAO from 'dao/EthereumDAO'

export const EXCHANGE_ORDERS = 'exchange/ORDERS'

export const search = (symbol: string, isBuy: boolean) => async dispatch => {
  const isSell = !isBuy

  let limitPromise
  let accountBalancePromise

  if ((symbol === ETH && isBuy) || (symbol === LHT && isSell)) {
    symbol = ETH
    limitPromise = exchangeDAO.getETHBalance()
    accountBalancePromise = exchangeDAO.getAccountAssetBalance()
  } else if ((symbol === ETH && isSell) || (symbol === LHT && isBuy)) {
    symbol = LHT
    limitPromise = exchangeDAO.getAssetBalance()
    accountBalancePromise = ethereumDAO.getAccountBalance()
  }

  const [limit, buyPrice, sellPrice, accountBalance] = await Promise.all([
    limitPromise,
    exchangeDAO.getBuyPrice(),
    exchangeDAO.getSellPrice(),
    accountBalancePromise,
  ])

  const order = new ExchangeOrderModel({
    limit,
    description: 'Auto Exchanger',
    symbol,
    isBuy,
    buyPrice,
    sellPrice,
    accountBalance,
  })

  dispatch({ type: EXCHANGE_ORDERS, orders: new Immutable.List([order]) })
}

export const exchange = (order: ExchangeOrderModel, amount: BigNumber) => async () => {
  try {
    if (order.isBuy()) {
      await exchangeDAO.buy(amount, order.sellPrice())
    } else {
      await exchangeDAO.sell(amount, order.buyPrice())
    }
  } catch (e) {
    // no rollback
  }
}
