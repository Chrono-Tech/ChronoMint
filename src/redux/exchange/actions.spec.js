import MainWalletModel from 'models/Wallet/MainWalletModel'
import ExchangesCollection from 'models/exchange/TokensCollection'
import BigNumber from 'bignumber.js'
import { WALLET_ALLOWANCE, WALLET_BALANCE, mainTransfer } from 'redux/mainWallet/actions'
import Immutable from 'immutable'
import networkService from 'Login/redux/network/actions'
import exchangeService from 'services/ExchangeService'
import { DUCK_SESSION } from 'redux/session/actions'
import { accounts, mockStore } from 'specsInit'
import ExchangeOrderModel from 'models/exchange/ExchangeOrderModel'
import ExchangeModel from 'models/exchange/ExchangeModel'
import * as a from './actions'

let store
const mock = new Immutable.Map({
  [DUCK_SESSION]: {
    account: accounts[0],
  },
  [a.DUCK_EXCHANGE]: new ExchangeModel({}),
})

describe('Exchange tests', () => {
  store = mockStore(mock)
  networkService.connectStore(store)

  let tokens = null
  let exchange = null
  it('should get tokens', async (done: Function) => {
    await store.dispatch(a.getTokenList())
    const actions = store.getActions()
    expect(actions[0].type).toEqual(a.EXCHANGE_GET_TOKENS_LIST_START)
    expect(actions[1].type).toEqual(a.EXCHANGE_GET_TOKENS_LIST_FINISH)
    expect(actions[1].tokens.size()).toEqual(2)
    tokens = actions[1].tokens
    done()
  })

  it.skip('should get exchange data', async (done: Function) => {
    await store.dispatch(a.getExchange())
    expect(await store.dispatch(a.getExchange())).toThrow()
    const actions = store.getActions()
    expect(actions[0].type).toEqual(a.EXCHANGE_GET_DATA_START)
    expect(actions[1].type).toEqual(a.EXCHANGE_GET_TOKENS_LIST_START)
    expect(actions[2].type).toEqual(a.EXCHANGE_GET_TOKENS_LIST_FINISH)
    expect(actions[2].tokens.size()).toEqual(2)
    expect(actions[3].type).toEqual(a.EXCHANGE_SET_PAGES_COUNT)
    expect(actions[3].count).toEqual(0)
    expect(actions[4].type).toEqual(a.EXCHANGE_GET_OWNERS_EXCHANGES_START)
    expect(actions[5].type).toEqual(a.EXCHANGE_MIDDLEWARE_DISCONNECTED)
    expect(actions[6].type).toEqual(a.EXCHANGE_GET_DATA_FINISH)
    expect(actions[7].type).toEqual(a.EXCHANGE_EXCHANGES_LIST_GETTING_START)
    done()
  })

  it('should get exchanges for owner', async (done: Function) => {
    const testMock = mock.set(a.DUCK_EXCHANGE, new ExchangeModel({ tokens }))
    store = mockStore(testMock)
    networkService.connectStore(store)
    await store.dispatch(a.getExchangesForOwner())
    const actions = store.getActions()
    expect(actions[0].type).toEqual(a.EXCHANGE_GET_OWNERS_EXCHANGES_START)
    expect(actions[1].type).toEqual(a.EXCHANGE_GET_OWNERS_EXCHANGES_FINISH)
    expect(actions[1].exchanges.size()).toEqual(jasmine.any(Number))
    done()
  })

  it('should create createExchange', async (done: Function) => {
    const testMock = mock.set(
      a.DUCK_EXCHANGE,
      new ExchangeModel({
        tokens,
        showFilter: false,
      })
    )
    store = mockStore(testMock)
    networkService.connectStore(store)

    const newExchange = new ExchangeOrderModel({
      buyPrice: new BigNumber(1),
      sellPrice: new BigNumber(1),
      symbol: 'TIME',
    })
    exchangeService.subscribeToCreateExchange(accounts[0])
    await exchangeService.on('ExchangeCreated', async (tx: Object) => {
      expect({
        buyPrice: tx.args.buyPrice,
        sellPrice: tx.args.sellPrice,
        symbol: tx.args.symbol,
      }).toMatchSnapshot()

      exchange = new ExchangeOrderModel({
        address: tx.args.exchange,
        symbol: tx.args.symbol,
      })
      done()
    })
    await store.dispatch(a.createExchange(newExchange))
  })

  it('should get allowance for token', async (done: Function) => {
    store.clearActions()
    await store.dispatch(a.getTokensAllowance(exchange))
    const actions = store.getActions()
    expect(actions[0].type).toEqual(WALLET_ALLOWANCE)
    expect(actions[0].token.allowance(exchange.address)).toEqual(new BigNumber(0))
    done()
  })

  it('should get exchanges count', async (done: Function) => {
    store.clearActions()
    await store.dispatch(a.getExchangesCount())
    const actions = store.getActions()
    expect(actions[0].type).toEqual(a.EXCHANGE_SET_PAGES_COUNT)
    expect(actions[0].count).toEqual(jasmine.any(Number))
    done()
  })

  it('should get next page', async (done: Function) => {
    store.clearActions()
    await store.dispatch(a.getNextPage())
    const actions = store.getActions()
    expect(actions[0].type).toEqual(a.EXCHANGE_EXCHANGES_LIST_GETTING_START)
    expect(actions[1].type).toEqual(a.EXCHANGE_EXCHANGES_LIST_GETTING_FINISH)
    expect(actions[1].exchanges.size()).toEqual(jasmine.any(Number))
    done()
  })

  it('should update exchange in redux', async (done: Function) => {
    const testMock = mock.set(
      a.DUCK_EXCHANGE,
      new ExchangeModel({
        tokens,
        exchanges: new ExchangesCollection().add(exchange),
        exchangesForOwner: new ExchangesCollection().add(exchange),
      })
    )
    store = mockStore(testMock)
    networkService.connectStore(store)

    await store.dispatch(a.updateExchange(exchange))
    const actions = store.getActions()
    expect(actions[0].type).toEqual(a.EXCHANGE_UPDATE_FOR_OWNER)
    expect(actions[1].type).toEqual(a.EXCHANGE_UPDATE)
    done()
  })

  it.skip('should withdraw from exchange', async (done: Function) => {
    store.clearActions()
    const token = tokens.getBySymbol('TIME')
    const address = exchange.address()

    await store.dispatch(mainTransfer(token, '10', address))

    const testMock = mock.set(a.DUCK_EXCHANGE, new ExchangeModel({ tokens }))
    store = mockStore(testMock)
    networkService.connectStore(store)
    await store.dispatch(a.getExchangesForOwner())

    const wallet = new MainWalletModel({
      address: accounts[0],
      tokens: new Immutable.Map({ TIME: token }),
    })
    exchangeService.subscribeToExchange(address)
    await exchangeService.on('WithdrawTokens', async tx => {
      expect(tx.exchange).toEqual(address)
      done()
    })
    await store.dispatch(a.withdrawFromExchange(exchange, wallet, '10', 'TIME'))
  })

  it('should get exchange from state', (done: Function) => {
    let state = new ExchangeModel({
      exchanges: new ExchangesCollection().add(exchange),
      exchangesForOwner: new ExchangesCollection().add(exchange),
    })
    const addresss = exchange.address()
    let exchangeFromState = a.getExchangeFromState(state, addresss)
    expect(exchangeFromState).toEqual(exchange)
    exchangeFromState = a.getExchangeFromState(state.exchanges(new ExchangesCollection()), addresss)
    expect(exchangeFromState).toEqual(exchange)
    done()
  })

  it('should make watchers for exchange', async (done: Function) => {
    try {
      await store.dispatch(a.watchExchanges())
      done()
    } catch (e) {
      done.fail()
    }
  })
})
