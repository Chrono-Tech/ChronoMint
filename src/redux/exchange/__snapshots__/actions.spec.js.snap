// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange tests should create createExchange 1`] = `
{
  "symbol": "0x54494d4500000000000000000000000000000000000000000000000000000000"
}
`;

exports[`Exchange tests should get exchange data 1`] = `
[
  {
    "type": "exchange/GET_DATA_START"
  },
  {
    "count": 0,
    "type": "exchange/EXCHANGE_SET_PAGES_COUNT"
  },
  {
    "type": "exchange/EXCHANGE_MIDDLEWARE_DISCONNECTED"
  },
  {
    "type": "exchange/GET_DATA_FINISH"
  },
  {
    "type": "exchange/EXCHANGE_GET_OWNERS_EXCHANGES_START"
  },
  {
    "type": "exchange/EXCHANGE_EXCHANGES_LIST_GETTING_START"
  }
]
`;
